rfcs/0007-custom-event-metadata-from-annotations/README.md

```diff
@@ -0,0 +1,99 @@
+* When implementing integrations, users often wish to embed various pieces of information in Flux objects that
+would help them identify entities or resources that are part of their integrations, e.g. a Deployment in the
+GitHub API.
+
+### Goals
+
+Provide a method for Flux users to embed custom/static fields in their Flux objects
+and have those fields sent to notification-controller as metadata on the event payload.
+
+### Non-Goals
+
+In this proposal we **do not** aim to provide a method for Flux users to send etcd-indexed custom metadata
+fields from Flux objects in events to notification-controller, most specifically labels. By design an event
+already contains enough identification information to locate the associated Flux object inside the cluster,
+and this proposal aims to provide a method for sending information that is relevant outside the cluster.
+Flux does not wish to incentive practices that are impactful to clusters without a strong reason or benefit.
+
+## Proposal
+
+When sending events about Flux objects, we propose sending annotation keys prefixed with a well-defined
+API Group in addition to the specific metadata fields already sent by the controller to implement Flux-specifc
+functionality.
+
+### User Stories
+
+* https://github.com/fluxcd/helm-controller/issues/680
+* https://github.com/fluxcd/helm-controller/pull/682#issuecomment-2079547110
+
+### Alternatives
+
+An alternative for extracting custom metadata fields from Flux objects to send on related events
+is defining `.spec` APIs for such, e.g. like `.spec.eventMetadata` available in the Alert API.
+This alternative is not great because:
+
+* Such APIs would be fairly redundant with the well-known Kubernetes annotations.
+* Technically speaking, it is much easier to implement an alternative where the
+fields storing the custom metadata are already available across all the Flux objects
+and can therefore be uniformly extracted without introducing a new API (which would
+be a sizable change across ~10 Flux APIs).
+
+## Design Details
+
+For the API Group used to extract the custom metadata fields from object annotations
+and accept those fields on the notification-controller side, we have a couple of alternatives:
+
+* `notification.toolkit.fluxcd.io`. This alternative emphasizes the close relationship of the feature
+with notification-controller and does not introduce a new API group.
+* `event.toolkit.fluxcd.io`. This alternative decouples the feature from notification-controller
+and brings the concept closer to Kubernetes Events. In fact, our implementation builds on the
+`EventRecorder` interface from `k8s.io/client-go/tools/record`, hijacking the `AnnotatedEventf()`
+logic to send events not only to Kubernetes but also notification-controller. This would, however,
+introduce a new Flux API Group.
+
+Chosen alternative: ?
```

Comment:
Please comment

Reply:
+1 for `event.toolkit.fluxcd.io`, the metadata will be present on Kubernetes events issued by Flux, so other tools can act on those based on the annotations, it's not only about notification-controller.

Reply:
But remember that n-c only accepts metadata fields prefixed specifically by the API Group of the involved object. If we want the annotations to be sent to Kubernetes with this new API Group we need to add an exception for it in n-c (which is not a big deal, we can probably do that).

Reply:
We can also keep this new API Group only on the event sent to Kubernetes and replace it with the API Group of the involved object. What do you think? @darkowlzz @stefanprodan

Reply:
No I think it's best to add the group to NC logic.

Reply:
thanks for the comments, changed the chosen alternative to `event.toolkit.fluxcd.io`